File: app/src/main/java/com/tasomaniac/devwidget/main/MainModule.java
package com.tasomaniac.devwidget.main;

import android.arch.lifecycle.ViewModel;
import android.support.v4.app.FragmentActivity;

import com.tasomaniac.devwidget.LifecycleScopeModule;
import com.tasomaniac.devwidget.ViewModelKey;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module(includes = LifecycleScopeModule.class)
public interface MainModule {

    @Binds
    FragmentActivity fragmentActivity(MainActivity activity);

    @Binds
    @IntoMap
    @ViewModelKey(MainModel.class)
    ViewModel mainModel(MainModel mainModel);

}

File: main/src/main/kotlin/com/tasomaniac/devwidget/main/MainModule.kt
package com.tasomaniac.devwidget.main

import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.ViewModel
import com.tasomaniac.devwidget.LifecycleScopeModule
import com.tasomaniac.devwidget.ViewModelKey
import com.tasomaniac.devwidget.ViewModelProviderModule
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module(
    includes = [
        LifecycleScopeModule::class,
        ViewModelProviderModule::class
    ]
)
internal interface MainModule {

    @Binds
    fun fragmentActivity(activity: MainActivity): FragmentActivity

    @Binds
    @IntoMap
    @ViewModelKey(MainModel::class)
    fun mainModel(mainModel: MainModel): ViewModel
}
